app/lib/primer/classify/functional_colors.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+module Primer
+  class Classify
+    # https://primer-css-git-mkt-color-modes-docs-primer.vercel.app/css/support/v16-migration
+    class FunctionalColors
+      FUNCTIONAL_TEXT_OPTIONS = {
+        primary: :text_primary,
+        secondary: :text_secondary,
+        tertiary: :text_tertiary,
+        link: :text_link,
+        success: :text_success,
+        warning: :text_warning,
+        danger: :text_danger,
+        white: :text_white,
+        inverse: :text_inverse
+      }.freeze
+
+      TEXT_COLOR_MAPPINGS = {
+        gray_dark: FUNCTIONAL_TEXT_OPTIONS[:primary],
+        gray: FUNCTIONAL_TEXT_OPTIONS[:secondary],
+        gray_light: FUNCTIONAL_TEXT_OPTIONS[:tertiary],
+        blue: FUNCTIONAL_TEXT_OPTIONS[:link],
+        green: FUNCTIONAL_TEXT_OPTIONS[:success],
+        yellow: FUNCTIONAL_TEXT_OPTIONS[:warning],
+        red: FUNCTIONAL_TEXT_OPTIONS[:danger],
+        white: FUNCTIONAL_TEXT_OPTIONS[:white],
+        # still unsure what will happen with these colors
+        black: nil,
+        orange: nil,
+        orange_light: nil,
+        purple: nil,
+        pink: nil,
+        inherit: nil
+      }.freeze
+
+      TEXT_OPTIONS = [
+        :icon_primary,
+        :icon_secondary,
+        :icon_tertiary,
+        :icon_info,
+        :icon_success,
+        :icon_warning,
+        :icon_danger,
+        *FUNCTIONAL_TEXT_OPTIONS.values
+      ].freeze
+      DEPRECATED_TEXT_OPTIONS = TEXT_COLOR_MAPPINGS.keys.freeze
+
+      class << self
+        def text_color(val)
+          # the value is a functional color
+          return "color-#{val.to_s.dasherize}" if ends_with_number?(val) || TEXT_OPTIONS.include?(val)
+          # if the app still allows non functional colors
+          return "text-#{val.to_s.dasherize}" unless force_functional_colors?
+
+          if TEXT_COLOR_MAPPINGS.key?(val)
+            functional_color = TEXT_COLOR_MAPPINGS[val]
+            # colors without functional mapping stay the same
+            return "text-#{val.to_s.dasherize}" if functional_color.blank?
+
+            ActiveSupport::Deprecation.warn("Color #{val} is deprecated. Please use #{functional_color} instead.") unless Rails.env.production? || silence_color_deprecations?
+
+            return "color-#{functional_color.to_s.dasherize}"
+          end
+
+          raise ArgumentError, "Color #{val} does not exist."
+        end
+
+        def ends_with_number?(val)
+          char_code = val[-1].ord
+          char_code >= 48 && char_code <= 57
+        end
+
+        def force_functional_colors?
+          Rails.application.config.primer_view_components.force_functional_colors
```

Comment:
I wonder if it'd be better for consumers to be able to opt in at the component level instead of at the app level. For example, in dotcom, it'd be nice to able to opt in for components instead of having to flip a switch for the app. Maybe I'm misunderstanding the goal here?

Reply:
hmm I guess that'd make it a lot more complex, since `Classify` is used by **all** components.

Reply:
I wonder if we might want to find a way to run this code in tests and then rewrite the values in the codebase, vs. converting these on the fly 🤔

Reply:
I wanted to avoid a breaking change here since they must be extensively used in dotcom. The deprecation warning should help us identify the old uses so we can update it as we go